### ISSUE TITLE
How can we get the  ['flow_feat']?

### ISSUE THREAD

**Issue by YangErNiu** [monjurulkarim/risky_object#12]
Dear Sir,
I noticed that there are ['feature'],['detection'],['vid_id'],['toa'] in the feat_extract.py, but there is no extraction method for [flow_feat]. [flow_feat] What is it? According to the thesis, we can know that ['feature'] already contains global feature variables and object feature variables, why do we need to input ['flow_feat'] into RiskyObject? 
@monjurulkarim

**Comment by monjurulkarim** [comment]
@YangErNiu 
['flow_feat'] is the optical flow which we extracted using[ RAFT](https://github.com/princeton-vl/RAFT). RAFT generates a flow video from which we extract the frames, treating them similarly to RGB images.
Each of the feature files contain the following information:

> **feature:** Extracted 2048-dim appearance features with shape (100, 31, 2048). It contains frame-level feature with shape (100, 1, 2048) and 19 box-level features with shape (100, 30, 2048)
> **detection:** Detected bounding boxes with shape (100, 30, 6), where the last dim denotes (object_tracking_id, x1, y1, x2, y2, cls), where cls =0 represents non risky and cls = 1 represents risky object.)
> **vid_id** : The video name for current feature file
> **toa:** time of accident when actual crash begins
> **flow_feat** : extracted the flow feature in the same format as the appearance feature

**Comment by YangErNiu** [comment]
I am sorry that I am not a native English speaker and a newcomer to this field. Therefore, when reading the paper and Fig.1 (AM-Net framework), I had some confusion, thinking that it was only necessary to input optical flow frames to extract feature vectors.🥲

Anyway,
I noticed that in the 'feat_extract.py' file, there is extraction for ['feature'] but not for ['flow_feat'].
Therefore, according to your answer, I need to input the original video frame and the optical flow frame generated by RAFT to get ['feature'],['flow_feat'].
So I just need to modify the feat_extract file to input the original video frames and optical flow frames, and perform the same operations on the optical flow frames with the featureExt(image) to get a file in the same format as the ROL CNN Feature. Or I can just copy a copy of feat_extract.py to change the input images to the optical images,then save the column name as ['flow_feat'], and concatenate two npz files to get the final npz file.
If I understand correctly, please give me a positive answer. Or let me know how you did it?Thank you very much. @monjurulkarim

**Comment by monjurulkarim** [comment]
@YangErNiu 
You're correct; we exclusively utilize optical flow features as input and do not incorporate appearance features. However, it's worth noting that our dataset does include appearance feature information should you wish to incorporate it.

The 'feat_extract.py' file is responsible for feature extraction. It accepts images (representing appearance) as input and generates a feature vector of shape (100, 31, 2048). As mentioned previously, we convert optical flow videos into images and subsequently perform feature extraction using the same 'feat_extract.py' file, resulting in another feature vector of shape (100, 31, 2048) containing the extracted flow feature. Upon extracting both appearance and optical flow features, we concatenate them to create the final feature file in *.npz format, with the following column names: [feature, detection, vid_id, toa, flow_feat]. Here, [feature] corresponds to appearance feature (unused in our method), while [flow_feat] represents the extracted optical flow feature.

**Comment by YangErNiu** [comment]
@monjurulkarim 
Thank you for your reply.
Based on the fact that the appearance features will not be used, after reading the paper again, I came to the conclusion that the ROL dataset was retained ['feature'] because it was previously used for ablation experiments.
But what makes me confused is that in demo.py, ['feature'] is input to the RiskyObject class. According to the comment in RiskyObject foward in the model.py : '''param x (batchsize, nFrames, 1+maxBox, Xdim), param y (batchsize, nFrames, maxBox, 6) '''.
It naturally occurred to me that x corresponds to ['feature'] and y corresponds to ['detection']. When I zeroed all the values in ['feature'] , I input demo.py again and find that the output is the same as the output without zeroing. I am sure that ['feature'] will not affect the result.

At the same time, there is a doubt that if the ['feature'] is only left for the convenience of ablation experiment, why does it need to be input into the model when running the demo.py? What is the role of param x? (The name of param 'y' is not too hard to understand, but I think it would be clear to name it 'detection'.)

But I feel that my understanding seems to be completely wrong, if you can take some time to explain it, I would be very grateful!

**Comment by nkamod** [comment]
I was trying to experiment with training the model with just appearance features instead of flow feature by replacing following line.
```python
#losses, all_outputs, all_labels = model(batch_xs, batch_det, batch_toas, batch_flow)
losses, all_outputs, all_labels = model(batch_xs, batch_det, batch_toas, batch_xs)
``` 

As you said, the final code, doesn't seem to use appearance features for training, except for getting shape of the data in some places.

The experiment with just appearance features, seems like performing the same. What's your take on it? Am I missing something from the paper? Computing optical flow is both computationally expensive and time consuming task. If the model trained on appearance features performs similar, compared to model trained on flow features, what should be the right way for the approach?

These are the results I received on testing - 

Checkpoints loaded successfully
Computing.........
AUC : 0.8452
AP : 0.5472
